### PR TITLE
Implement md.{DataFrame,Series,GroupBy}.apply

### LINKED ISSUE #568: Implement DataFrame groupby
**Is your feature request related to a problem? Please describe.**
Implement `groupby` operator, its return value is a new DataFrame type `DataFrameGroupBy` and users can do computation on it, such as reduce function or `apply`, `transform`, etc.

### LINKED ISSUE #1034: Implement md.{DataFrame,Series,GroupBy}.apply
Implements apply function separately for data frames, series and groupby results.

Applying dicts / lists may need to rely on #1035 
Full support of groupby() may rely on #1036

### PR DESCRIPTION
**Pull request by wjsi**
Implement md.{DataFrame,Series,GroupBy}.apply and md.GroupBy.transform for data frames. To support SeriesGroupBy.apply, a rough version of Series.groupby is also added.

Applying by lists / dicts not implemented as support for {DataFrame,Series}.aggregate is missing. groupby(..., sort=True, as_index=False).apply(func) is not implemented as groups are not sorted.

This PR fixes #1034
This PR resolves #568 

As cloudpickle is depended, this PR cannot be backported.

### CODE REVIEW

File: mars/dataframe/base/apply.py [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+                    new_shape = c.shape
+                else:
+                    new_shape = [np.nan, np.nan]
+                    new_shape[1 - axis] = c.shape[1 - axis]
+
+                new_dtypes = out_df.dtypes[c.dtypes.keys()]
+
+                new_op = op.copy().reset_key()
+                chunks.append(new_op.new_chunk([c], shape=tuple(new_shape), index=c.index, dtypes=new_dtypes,
+                                               index_value=c.index_value, columns_value=c.columns_value))
```
Comment (qinxuye) [inline mars/dataframe/base/apply.py]: `columns_value` seems incorrect.

File: mars/dataframe/groupby/apply.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+                index_value = parse_index(pd.RangeIndex(-1))
+
+            if isinstance(infer_df, pd.DataFrame):
+                object_type = object_type or ObjectType.dataframe
+                new_dtypes = new_dtypes or infer_df.dtypes
+            elif isinstance(infer_df, pd.Series):
+                object_type = object_type or ObjectType.series
+                new_dtypes = new_dtypes or infer_df.dtype
+            else:
+                object_type = ObjectType.series
```
Comment (qinxuye) [inline mars/dataframe/groupby/apply.py]: `ObjectType` could be scalar which is actually a tensor.

Comment (wjsi) [reply mars/dataframe/groupby/apply.py]: When the returned value of the provided function is a scalar, groupby().apply() actually returns a series.

File: mars/dataframe/base/apply.py [hunk 3]
```diff
@@ -102,21 +102,29 @@ def tile(cls, op):
...
+                    new_index_value, new_columns_value = c.index_value, c.columns_value
                 else:
                     new_shape = [np.nan, np.nan]
                     new_shape[1 - axis] = c.shape[1 - axis]
+                    if axis == 0:
+                        new_index_value = parse_index(pd.RangeIndex(-1))
+                        new_columns_value = c.columns_value
+                    else:
+                        new_index_value = c.index_value
+                        new_columns_value = parse_index(pd.RangeIndex(-1))
```
Comment (qinxuye) [inline mars/dataframe/base/apply.py]: When axis == 1, due to `rechunk`, maybe the `columns_value` is just `out_df.columns_value`?

File: mars/serialize/protos/operand.proto [hunk 4]
```diff
@@ -348,8 +348,8 @@ message OperandDef {
         STORE_COO = 1004;
 
         // apply
-        DATAFRAME_APPLY = 1101;
-        SERIES_APPLY = 1102;
+        DATAFRAME_APPLY_TRANSFORM = 1101;
```
Comment (qinxuye) [inline mars/serialize/protos/operand.proto]: IMO, it's better to define apply and transform separately.

**Pull request merged**